upup/pkg/fi/cloudup/containerd.go

```diff
@@ -0,0 +1,173 @@
+		if dv != "" {
+			var err error
+			containerdAssetUrl, err = findDockerVersionUrl(arch, dv)
+			if err != nil {
+				return "", "", err
+			}
+			containerdAssetHash, err = findDockerVersionHash(arch, dv)
+			if err != nil {
+				return "", "", err
+			}
+			println(dv)
+		} else {
+			return "", "", fmt.Errorf("unknown url and hash for containerd version: %s - %s", arch, version)
+		}
+	}
+
+	return containerdAssetUrl, containerdAssetHash, nil
+}
+
+func findContainerdVersionUrl(arch architectures.Architecture, version string) (string, error) {
+	sv, err := semver.ParseTolerant(version)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse version string: %q", version)
+	}
+	if sv.LT(semver.MustParse("1.3.4")) {
+		return "", fmt.Errorf("unsupported legacy containerd version: %q", version)
+	}
+
+	var u string
+	switch arch {
+	case architectures.ArchitectureAmd64:
+		if sv.GTE(semver.MustParse("1.4.0")) {
+			u = fmt.Sprintf(containerdVersionUrlAmd64, version, version)
+		} else {
+			u = fmt.Sprintf(containerdLegacyUrlAmd64, version)
+		}
+	case architectures.ArchitectureArm64:
+		// For now there are only official AMD64 builds, using Default Docker version instead
+		if findAllContainerdHashesAmd64()[version] != "" {
+			u = fmt.Sprintf(dockerVersionUrlArm64, findAllContainerdDockerMappings()[containerdFallbackVersion])
+		}
+	default:
+		return "", fmt.Errorf("unknown arch: %q", arch)
+	}
+
+	if u == "" {
+		return "", fmt.Errorf("unknown url for containerd version: %s - %s", arch, version)
+	}
+
+	return u, nil
+}
+
+func findContainerdVersionHash(arch architectures.Architecture, version string) (string, error) {
+	sv, err := semver.ParseTolerant(version)
+	if err != nil {
+		return "", fmt.Errorf("unable to parse version string: %q", version)
+	}
+	if sv.LT(semver.MustParse("1.3.4")) {
+		return "", fmt.Errorf("unsupported legacy containerd version: %q", version)
+	}
+
+	var h string
+	switch arch {
+	case architectures.ArchitectureAmd64:
+		h = findAllContainerdHashesAmd64()[version]
+	case architectures.ArchitectureArm64:
+		// For now there are only official AMD64 builds, using Default Docker version instead
+		if findAllContainerdHashesAmd64()[version] != "" {
+			h = findAllDockerHashesArm64()[findAllContainerdDockerMappings()[containerdFallbackVersion]]
+		}
+	default:
+		return "", fmt.Errorf("unknown arch: %q", arch)
+	}
+
+	if h == "" {
+		return "", fmt.Errorf("unknown hash for containerd version: %s - %s", arch, version)
+	}
+
+	return h, nil
+}
+
+func findAllContainerdHashesAmd64() map[string]string {
+	hashes := map[string]string{
+		"1.3.4": "4616971c3ad21c24f2f2320fa1c085577a91032a068dd56a41c7c4b71a458087",
+		"1.4.0": "b379f29417efd583f77e095173d4d0bd6bb001f0081b2a63d152ee7aef653ce1",
+		"1.4.1": "757efb93a4f3161efc447a943317503d8a7ded5cb4cc0cba3f3318d7ce1542ed",
+	}
+
+	return hashes
+}
+
+func findAllContainerdDockerMappings() map[string]string {
+	versions := map[string]string{
```

Comment:
So:
* drops 1.2.4
* adds 1.2.6 and 1.2.12
* changes 1.2.10 from containerd to docker
* changes 1.2.13 from docker 19.03.11 to 19.03.12

Shouldn't at least the dropping of 1.2.4 be mentioned in release notes?

Reply:
Containerd support is a bit of a puzzle here. Will have to think about it.

Reply:
I would be okay with only moving the newer versions of containerd to assets phase.

Reply:
The problem comes mostly from packaging bugs in containerd and package URL changes and lack of ARM builds...
So, to the point:
* drops 1.2.4 - adding back pretty easy to do
* adds 1.2.6 and 1.2.12 - it is not a big deal, added them just because it was easy to map to Docker packages
* changes 1.2.10 from containerd to docker - intended, shouldn't make any difference
* changes 1.2.13 from docker 19.03.11 to 19.03.12 - reverted this